### ISSUE TITLE
Partitions are not locked during DROP TABLE

### ISSUE THREAD

**Issue by hlinnaka** [greenplum-db/gpdb#5919]
A `DROP TABLE` command on a partitioned table doesn't acquire locks on the partitions. That can lead to all kinds of bad things, if a partition is accessed concurrently.

For example:
In psql session A:
```
postgres=# create table partlockt (i int, t text) partition by range(i)
(start(1) end(3) every(1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'i' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "partlockt_1_prt_1" for table "partlockt"
NOTICE:  CREATE TABLE will create partition "partlockt_1_prt_2" for table "partlockt"
CREATE TABLE
postgres=# begin;
BEGIN
postgres=# drop table partlockt;
DROP TABLE
[leave the transaction open]
```
In psql session B:
```
postgres=# begin;
BEGIN
postgres=# insert into partlockt_1_prt_1 select 1 from generate_series(1, 10000000);
```
While that's running, switch back to session A:
```
postgres=# commit;
COMMIT
```
Now you get an error in session B (or some other weird behavior):
```
ERROR:  relation with OID 16951 does not exist
```

That wouldn't be so bad, but now I'm seeing this in session A, when I try to create a new session:
```
postgres=# create table foo (id int4);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
WARNING:  could not write block 8858 of base/12549/16477  (seg2 127.0.0.1:40002 pid=6600)
DETAIL:  Multiple failures --- write error might be permanent.
ERROR:  could not open file "base/12549/16477": No such file or directory  (seg2 127.0.0.1:40002 pid=6600)
CONTEXT:  writing block 8858 of relation base/12549/16477
```

The bottom line is that we really should hold a lock on each partition that's being dropped. We can't play fast and loose with that.

**Comment by BaiShaoqi** [comment]
Could I use the following code to deal with issue?  To make diff look small, I use ugly code here.

```
index af12801..83af4af 100644
--- a/src/backend/commands/tablecmds.c
+++ b/src/backend/commands/tablecmds.c
@@ -14941,7 +14941,8 @@ rel_needs_long_lock(Oid relid)
        relation_close(rel, NoLock);

        if (Gp_role == GP_ROLE_DISPATCH)
-               needs_lock = !rel_is_child_partition(relid);
+               ;
+               //needs_lock = !rel_is_child_partition(relid);
        else
        {
                Relation inhrel;
```

By the way, there is a comment saying,  so I think `Partitions are not locked during DROP TABLE` was intended, although it results to wrong behavior as you have found in this issue.
```
		/*
		 * GPDB: If this was a partition, or some other object for which
		 * we are careful to always lock the parent object, we don't need
		 * to keep the lock on the sub-object. This helps to keep the lock
		 * table size in check, if you e.g. drop a table with thousands
		 * of partitions. It would be unpleasent if you could not drop
		 * such a table because the lock table runs out of space.
```

**Comment by khuddlefish** [comment]
The commit that was fixing this issue was reverted due to a test failure. Reopening issue
